packages/snaps-controllers/src/snaps/SnapController.ts

```diff
@@ -968,6 +995,96 @@ export class SnapController extends BaseController<
     );
   }
 
+  #handlePreinstalledSnaps(preinstalledSnaps: PreinstalledSnap[]) {
+    for (const { snapId, manifest, files, removable } of preinstalledSnaps) {
+      const existingSnap = this.get(snapId);
+      const isAlreadyInstalled = existingSnap !== undefined;
+      const isUpdate =
+        isAlreadyInstalled && gtVersion(manifest.version, existingSnap.version);
+
+      // Disallow downgrades and overwriting non preinstalled snaps
+      if (
+        isAlreadyInstalled &&
+        (!isUpdate || existingSnap.preinstalled !== true)
+      ) {
+        continue;
+      }
+
+      const manifestFile = new VirtualFile<SnapManifest>({
+        path: NpmSnapFileNames.Manifest,
+        value: JSON.stringify(manifest),
+        result: manifest,
+      });
+
+      const virtualFiles = files.map(
+        ({ path, value }) => new VirtualFile({ value, path }),
+      );
+      const { filePath, iconPath } = manifest.source.location.npm;
+      const sourceCode = virtualFiles.find((file) => file.path === filePath);
+      const svgIcon = iconPath
+        ? virtualFiles.find((file) => file.path === iconPath)
+        : undefined;
+
+      assert(sourceCode, 'Source code not provided for preinstalled snap');
+
+      assert(
+        manifest.source.files === undefined,
+        'Auxiliary files are not currently supported for preinstalled snaps',
+      );
+
+      assert(
+        manifest.source.locales === undefined,
+        'Localization files are not currently supported for preinstalled snaps',
```

Comment:
This may be something we want to support?

Reply:
Yeah, problem is that some of this is async, at least for auxiliary files.

Reply:
I added support for localization files at least. We can deal with auxiliary files when it becomes a problem IMO.